### PR TITLE
Add support for getting debugger data tips through LSP.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/EditorFeatures/CSharpTest/Debugging/DataTipInfoGetterTests.cs [hunk 1]
```diff
@@ -2,60 +2,20 @@
...
 using Xunit;
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.Debugging;
 
 [UseExportProvider]
 [Trait(Traits.Feature, Traits.Features.DebuggingDataTips)]
-public class DataTipInfoGetterTests
+public sealed class DataTipInfoGetterTests : AbstractDataTipInfoGetterTests
 {
-    private static async Task TestAsync(string markup, string expectedText = null)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Debugging/DataTipInfoGetterTests.cs:25]: moved to base type to share with VB.

File: src/EditorFeatures/VisualBasicTest/Debugging/DataTipInfoGetterTests.vb [hunk 2]
```diff
@@ -2,57 +2,16 @@
...
-Imports Roslyn.Utilities
+Imports Microsoft.CodeAnalysis.Test.Utilities.Debugging
 
 Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.Debugging
-
     <[UseExportProvider]>
     <Trait(Traits.Feature, Traits.Features.DebuggingDataTips)>
-    Public Class DataTipInfoGetterTests
-
-        Private Shared Async Function TestNoDataTipAsync(input As XElement) As Task
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasicTest/Debugging/DataTipInfoGetterTests.vb:22]: removed and replaced with calls to shared base class methods.

File: src/LanguageServer/Protocol/Handler/DataTips/DataTipRangeHandler.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(TextDocumentPositionParams request)
+        => request.TextDocument;
+
+    public async Task<VSInternalDataTip?> HandleRequestAsync(TextDocumentPositionParams request, RequestContext context, CancellationToken cancellationToken)
+    {
+        var document = context.GetRequiredDocument();
+
+        var service = document.GetLanguageService<ILanguageDebugInfoService>();
+        if (service is null)
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Handler/DataTips/DataTipRangeHandler.cs:34]: Can this ever be null?  I think I saw a C#/VB/F# implementation

File: src/LanguageServer/Protocol/Protocol/Internal/VSInternalServerCapabilities.cs [hunk 4]
```diff
@@ -4,6 +4,7 @@
 
 namespace Roslyn.LanguageServer.Protocol
 {
+    using System.Runtime.Serialization;
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Protocol/Internal/VSInternalServerCapabilities.cs:7]: is this required?

File: src/LanguageServer/Protocol/Protocol/Internal/VSInternalDataTips.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    /// <summary>
+    /// Data tip range is a linq expression.
+    /// </summary>
+    LinqExpression = 0x1,
+}
+
+/// <summary>
+/// Class which represents debugger data tip response.
+/// </summary>
+[DataContract]
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/Protocol/Internal/VSInternalDataTips.cs:26]: remove - data contract attributes don't apply to STJ

File: src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+            var v = args.{|caret:|}Select(a => a.ToString()).Where(a => a.Length >= 0);
+            """;
+        await using var testLspServer = await CreateTestLspServerAsync(markup, mutatingLspWorkspace);
+
+        var caret = testLspServer.GetLocations("caret").Single();
+
+        var result = await RunAsync(testLspServer, caret);
+        Assert.NotNull(result);
+
+        AssertJsonEquals(new LSP.VSInternalDataTip
```
Comment (dibarbet) [inline src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs:46]: should we consider making `VSInternalDataTip` a record type so we could just equals these?

File: src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        {
+            DataTipTags = LSP.VSInternalDataTipTags.LinqExpression,
+            HoverRange = new LSP.Range
```
Comment (dibarbet) [inline src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs:49]: could consider annotating the text above with the expected output ranges

File: src/VisualStudio/Core/Def/Implementation/AbstractVsTextViewFilter.cs [hunk 8]
```diff
@@ -119,7 +119,10 @@ protected int GetDataTipTextImpl(ITextBuffer subjectBuffer, TextSpan[] pSpan, ou
                             var spanOpt = textSnapshot.TryGetSpan(textSpan);
                             if (spanOpt.HasValue)
                             {
-                                var dataTipInfo = languageDebugInfo.GetDataTipInfoAsync(document, spanOpt.Value.Start, cancellationToken).WaitAndGetResult(cancellationToken);
+                                // 'kind' is an lsp-only concept, so we don't want/need to include it here (especially
+                                // as it can be expensive to compute, and we don't want to block the UI thread).
+                                var dataTipInfo = languageDebugInfo.GetDataTipInfoAsync(
```
Comment (dibarbet) [inline src/VisualStudio/Core/Def/Implementation/AbstractVsTextViewFilter.cs:124]: if we're returning data tips from LSP, why are we computing it again here in a MEFish API?   Probably missing something about the design here

File: src/LanguageServer/Protocol/DefaultCapabilitiesProvider.cs [hunk 9]
```diff
@@ -129,6 +129,7 @@ private static VSInternalServerCapabilities GetVSServerCapabilities()
             {
                 ProjectContextProvider = true,
                 BreakableRangeProvider = true,
+                DataTipRangeProvider = true,
```
Comment (dibarbet) [inline src/LanguageServer/Protocol/DefaultCapabilitiesProvider.cs:132]: if this is expected to be active all the time in VS, it needs to be added here - https://github.com/dotnet/roslyn/blob/main/src/EditorFeatures/Core/LanguageServer/AlwaysActivateInProcLanguageClient.cs#L51

these capabilities only apply when we're in full LSP mode (feature flag, vscode).

File: src/LanguageServer/Protocol/Handler/DataTips/DataTipRangeHandler.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Handler/DataTips/DataTipRangeHandler.cs:34]: i wasn't certain what our rules were with LSP.  like how xaml works there.  i wwanted to be a little safe.

File: src/LanguageServer/Protocol/Protocol/Internal/VSInternalDataTips.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Protocol/Internal/VSInternalDataTips.cs]: done.

File: src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs:46]: sure!

File: src/LanguageServer/Protocol/Handler/DataTips/DataTipRangeHandler.cs (hunk 3 above)
Comment (dibarbet) [reply src/LanguageServer/Protocol/Handler/DataTips/DataTipRangeHandler.cs:34]: this is exported as `ExportCSharpVisualBasicStatelessLspService`, so it applies only to the main roslyn server (which supports C#,VB, and F#).  It won't be called for anything else

File: src/LanguageServer/Protocol/Protocol/Internal/VSInternalServerCapabilities.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/Protocol/Internal/VSInternalServerCapabilities.cs]: no.  removed.

File: src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs (hunk 7 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/ProtocolUnitTests/DataTips/DataTipRangeHandlerTests.cs:49]: i thought about it, but didn't care.  this is really just a single test to validate the data once through LSP..  the actual unit tests are at a lower level.

File: src/LanguageServer/Protocol/DefaultCapabilitiesProvider.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/LanguageServer/Protocol/DefaultCapabilitiesProvider.cs]: Moved.

File: src/VisualStudio/Core/Def/Implementation/AbstractVsTextViewFilter.cs (hunk 8 above)
Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/AbstractVsTextViewFilter.cs:124]: asnwered offline.  will support the old and new systems for a short period of time.

**Review by dibarbet — APPROVED** [review]

File: src/Features/Core/Portable/Debugging/DebugDataTipInfo.cs [hunk 10]
```diff
@@ -6,10 +6,19 @@
...
+internal enum DebugDataTipInfoKind
 {
-    public readonly TextSpan Span = span;
-    public readonly string Text = text;
+    None,
+    LinqExpression,
+}
+
+internal readonly record struct DebugDataTipInfo(
+    TextSpan Span, TextSpan ExpressionSpan, string? Text, DebugDataTipInfoKind Kind = DebugDataTipInfoKind.None)
```
Comment (dustincoleman) [inline src/Features/Core/Portable/Debugging/DebugDataTipInfo.cs:16]: >TextSpan Span [](http://example.com/codeflow?start=4&length=13)

Have we implemented this such that if the cursor in anywhere in Span, the same expression to evaluate will be produced?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/Debugging/DebugDataTipInfo.cs:16]: Yes. That should be the semantics of 'span'.

**Pull request merged**